upload/tests/views/test_combined_upload.py

```diff
@@ -0,0 +1,289 @@
+    )
+    assert response.status_code == 400
+    assert "flags" in response.json()
+
+
+def test_combined_upload_post(db, mocker):
+    mocker.patch.object(
+        CanDoCoverageUploadsPermission, "has_permission", return_value=True
+    )
+    presigned_put_mock = mocker.patch(
+        "services.archive.StorageService.create_presigned_put",
+        return_value="presigned put",
+    )
+    upload_task_mock = mocker.patch(
+        "upload.views.uploads.trigger_upload_task", return_value=True
+    )
+
+    repository = RepositoryFactory(
+        name="the_repo1", author__username="codecov", author__service="github"
+    )
+    commit = CommitFactory(repository=repository)
+    repository.save()
+    commit.save()
+
+    owner = repository.author
+    client = APIClient()
+    client.force_authenticate(user=owner)
+    repo_slug = f"{repository.author.username}::::{repository.name}"
+    url = reverse(
+        "new_upload.combined_upload",
+        args=[repository.author.service, repo_slug],
+    )
+    response = client.post(
+        url,
+        {
+            "branch": "branch",
+            "ci_service": "ci_service",
+            "ci_url": "ci_url",
+            "code": "code",
+            "commit_sha": commit.commitid,
+            "flags": ["flag1", "flag2"],
+            "job_code": "job_code",
+            "version": "version",
+        },
+        format="json",
+    )
+    response_json = response.json()
+    upload = ReportSession.objects.filter(
+        report__commit=commit,
+        report__code="code",
+        upload_extras={"format_version": "v1"},
+    ).first()
+    assert response.status_code == 201
+    assert all(
+        map(
+            lambda x: x in response_json.keys(),
+            ["external_id", "created_at", "raw_upload_location", "url"],
+        )
+    )
+    assert (
+        response_json.get("url")
+        == f"{settings.CODECOV_DASHBOARD_URL}/{repository.author.service}/{repository.author.username}/{repository.name}/commit/{commit.commitid}"
+    )
+
+    assert ReportSession.objects.filter(
+        report__commit=commit,
+        report__code="code",
+        upload_extras={"format_version": "v1"},
+    ).exists()
+    assert RepositoryFlag.objects.filter(
+        repository_id=repository.repoid, flag_name="flag1"
+    ).exists()
+    assert RepositoryFlag.objects.filter(
+        repository_id=repository.repoid, flag_name="flag2"
+    ).exists()
+    flag1 = RepositoryFlag.objects.filter(
+        repository_id=repository.repoid, flag_name="flag1"
+    ).first()
+    flag2 = RepositoryFlag.objects.filter(
+        repository_id=repository.repoid, flag_name="flag2"
+    ).first()
+    assert UploadFlagMembership.objects.filter(
+        report_session_id=upload.id, flag_id=flag1.id
+    ).exists()
+    assert UploadFlagMembership.objects.filter(
+        report_session_id=upload.id, flag_id=flag2.id
+    ).exists()
+    assert [flag for flag in upload.flags.all()] == [flag1, flag2]
```

Comment:
instead of these exists calls, I would just assert that the flags are properly attached:

```
assert {flag.name for flag in upload.flags.all()} == {"flag1", "flag2"}
```

We shouldn’t care about the implementation details of how the database manages its relations and joins. We care that our upload has the proper flags.